### PR TITLE
Add ClipVessel module

### PR DESCRIPTION
**Pull request by dmolony3**
Clip segmentation/model based on user provided centerlines and markup points

### CODE REVIEW

File: ClipVessel/ClipVessel.py [hunk 1]
```diff
@@ -0,0 +1,800 @@
...
+    capDisplacement = 0.0
+    surfaceCapper = vtkvmtkComputationalGeometry.vtkvmtkCapPolyData()
+    surfaceCapper.SetInputData(surface)
+    surfaceCapper.SetDisplacement(capDisplacement)
+    surfaceCapper.SetInPlaneDisplacement(capDisplacement)
+    surfaceCapper.Update()
+    surface = surfaceCapper.GetOutput()
+    return surface
+    
+  def preprocess(self, surfacePolyData, targetNumberOfPoints, decimationAggressiveness, subdivide):
```
Comment (lassoan) [inline ClipVessel/ClipVessel.py:379]: Duplicating code would significantly increase maintenance workload. Please call preprocess method of ExtractCenterline.py instead. You can do something like this:

```
import ExtractCenterline
extractCenterlineLogic = ExtractCenterline.ExtractCenterlineLogic()
extractCenterlineLogic.preprocess(...)
```

File: ClipVessel/ClipVessel.py [hunk 2]
```diff
@@ -0,0 +1,800 @@
...
+    normals.SetAutoOrientNormals(1)
+    normals.SetFlipNormals(0)
+    normals.SetConsistency(1)
+    normals.SplittingOff()
+    normals.Update()
+
+    return normals.GetOutput()
+
+  # Unclear if necessary
+  def extractNonManifoldEdges(self, polyData, nonManifoldEdgesPolyData=None):
```
Comment (lassoan) [inline ClipVessel/ClipVessel.py:441]: Call ExtractCenterline logic instead of duplicating code

File: ClipVessel/ClipVessel.py [hunk 3]
```diff
@@ -0,0 +1,800 @@
...
+    cutter.SetInputData(surface)
+    cutter.SetCutFunction(clipFunctionPlane)
+    
+    clippedSurface = vtk.vtkPolyData()
+    cutLines = vtk.vtkPolyData()
+    clipper.Update()
+    
+    surface.DeepCopy(clipper.GetOutput())
+    clippedSurface.DeepCopy(clipper.GetClippedOutput())
+    cutter.Update()
```
Comment (lassoan) [inline ClipVessel/ClipVessel.py:552]: No need for

```
filter.Update()
data=filter.GetOutput()
filter2.SetInputData(data)
```

instead you can simply write

```
filter2.SetInputConnection(filter.GetOutputPort())
```

File: ClipVessel/ClipVessel.py [hunk 4]
```diff
@@ -0,0 +1,800 @@
...
+
+  def __init__(self, parent):
+    ScriptedLoadableModule.__init__(self, parent)
+    self.parent.title = "Clip Vessel"
+    self.parent.categories = ["Vascular Modeling Toolkit"]
+    self.parent.dependencies = []
+    self.parent.contributors = ["David Molony (NGHS)", "Andras Lasso (PerkLab)"]
+    self.parent.helpText = """
+This module clips a surface model given a VMTK centerline and markups indicating where the model will be clipped. The first marker indicates the inlet. Optionally, the user can cap and add flow extensions.
+    <a href="https://github.com/vmtk/SlicerExtension-VMTK/">here</a>.
```
Comment (lassoan) [inline ClipVessel/ClipVessel.py:27]: add url to the module documentation (not just the repository) to make the documentation easier to find

File: ClipVessel/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,31 @@
+#-----------------------------------------------------------------------------
+set(MODULE_NAME ClipVessel)
+
+#-----------------------------------------------------------------------------
+set(MODULE_PYTHON_SCRIPTS
+  ${MODULE_NAME}.py
+  )
+
+set(MODULE_PYTHON_RESOURCES
+  Resources/Icons/${MODULE_NAME}.png
```
Comment (lassoan) [inline ClipVessel/CMakeLists.txt:10]: Please add a new icon. You can generate it with any AI image generator.

**Review by lassoan — COMMENTED** [review]
Thank you very much for your contribution, this will be very useful. I've just added a couple of comments that would be great if you could address before merging.

File: ClipVessel/CMakeLists.txt (hunk 5 above)
Comment (dmolony3) [reply ClipVessel/CMakeLists.txt:10]: Hi @lassoan, 

Sorry for the long time to address these issues. I have updated the module. Please let me know if there are any remaining issues.

**Comment by dmolony3** [comment]
Hi @lassoan,

Sorry for the long time to address these issues. I have updated the module. Please let me know if there are any remaining issues.

**Comment by lassoan** [comment]
@chir-set could you have a look at this pull request? If it works well for you and you think it could be useful then you could merge it.

**Comment by chir-set** [comment]
> then you could merge it.

I was about to merge and on last check, this error happens very often.

Traceback (most recent call last):
  File "/home/user/Documents/Slicer4/Custom_Ext/ClipVessel/ClipVessel/ClipVessel.py", line 281, in onApplyButton
    outputPolyData = self.logic.clipVessel(preprocessedPolyData, centerlinesModelNode, clipPointsMarkupsNode, cap, addFlowExtensions, extensionLength, extensionMode)
  File "/home/user/Documents/Slicer4/Custom_Ext/ClipVessel/ClipVessel/ClipVessel.py", line 743, in clipVessel
    num_regions = int(surface.GetPointData().GetArray("RegionId").GetRange()[1])
AttributeError: 'NoneType' object has no attribute 'GetRange'

There's some NULL check to understand and manage. I should probably be able to do that. However, I'm soon away for 2 weeks and somehow busy. I'll look into that in January.

**Pull request merged**